src/lib/rules/amp-validator/amp-validator.ts

```diff
@@ -0,0 +1,91 @@
+/**
+ * @fileoverview Validates the HTML of a page is AMP valid
+ */
+
+import * as amphtmlValidator from 'amphtml-validator';
+
+// The list of types depends on the events you want to capture.
+import { Category } from '../../enums/category';
+import { RuleContext } from '../../rule-context';
+import { IRule, IRuleBuilder } from '../../types';
+import { debug as d } from '../../utils/debug';
+
+const debug: debug.IDebugger = d(__filename);
+
+import { ITargetFetchEnd, IScanEnd } from '../../types';
+
+// ------------------------------------------------------------------------------
+// Public
+// ------------------------------------------------------------------------------
+
+const rule: IRuleBuilder = {
+    create(context: RuleContext): IRule {
+        const validPromise = amphtmlValidator.getInstance();
+        const errorsOnly = context.ruleOptions && context.ruleOptions['error-only'] || false;
+        let html;
+
+        const onTargetFetchEnd = (fetchEnd: ITargetFetchEnd) => {
+            const { response: { body: { content }, statusCode } } = fetchEnd;
+
+            if (statusCode !== 200 || !content) {
+                return;
+            }
+
+            html = content;
+        };
+
+        const onScanEnd = async (scanEnd: IScanEnd) => {
+            if (!html) {
+                debug('No valid content');
+
+                return;
+            }
+
+            const resource = scanEnd.resource;
+            const validator = await validPromise;
+            const result = validator.validateString(html);
+
+            if (result === 'PASS') {
+                debug('Content is AMP valid');
+
+                return;
+            }
+
+            for (let i = 0; i < result.errors.length; i++) {
+                const error = result.errors[i];
+                let message = error.message;
+
+                if (error.specUrl !== null) {
+                    message += ` (${error.specUrl})`;
+                }
+
+                // We ignore errors that are not 'ERROR' if user has configured the rule like that
+                if (errorsOnly && error.severity !== 'ERROR') {
+                    debug(`AMP error doesn't meet threshold for reporting`);
+                } else {
+                    await context.report(resource, null, message, null, { column: error.column, line: error.line });
+                }
+            }
+        };
+
+        return {
+            'scan::end': onScanEnd,
```

Comment:
Why is this event necessary? You already have the html in the event `targetfetch::end`

Reply:
My idea is to not block the other events and let them fire asap, similar to what we do in `html-checker`, `ssl-labs`, etc. In this case, the creation of a validator via `amphtmlValidator` returns a `Promise` in:

```js
const validPromise = amphtmlValidator.getInstance(); 
```
so I'm waiting for it in the `scan::end` event. I should probably move the creation to `targetfetch::end` or just combine everything on the same handler if we think it's best (personally I prefer to wait until the end to do the check).

Reply:
Ok! I think is better too do the check at the end

Reply:
I'll move the promise creation to the `targetfetch::end`. No point in creating it if there is an error in the process.

Reply:
ok!